plugins/optimization/src/main/java/org/qbicc/plugin/opt/ea/ConnectionGraph.java

```diff
@@ -0,0 +1,303 @@
+package org.qbicc.plugin.opt.ea;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.qbicc.graph.Call;
+import org.qbicc.graph.CheckCast;
+import org.qbicc.graph.InstanceFieldOf;
+import org.qbicc.graph.Load;
+import org.qbicc.graph.LocalVariable;
+import org.qbicc.graph.New;
+import org.qbicc.graph.Node;
+import org.qbicc.graph.ParameterValue;
+import org.qbicc.graph.PhiValue;
+import org.qbicc.graph.Value;
+import org.qbicc.graph.ValueHandle;
+
+final class ConnectionGraph {
+    // TODO Handle situations where a node has multiple points-to.
+    //      Even if a reference is potentially assigned multiple New nodes (e.g. branches), the refs are currently different.
+    private final Map<Node, Node> pointsToEdges = new HashMap<>(); // solid (P) edges
```

Comment:
Points-to are edges from references to objects referenced. The references are `ValueHandle` instances, e.g. `ReferenceHandle`, `StaticField`...etc. Referenced objects are normally `New` instances, but they can also be `ParameterValue` or `Call`.

Each method's connection graph tracks this during intra method analysis. This is used for multiple things:

1. The points-to edges are used to transfer information from callee methods to caller methods. For example, imagine that a method A creates a new `Boo` and assigns it to variable `x`. Then, method A sends `x` as argument to method B. If method B assigns the parameter to a static variable, the parameter reference is marked as global escape. However, method B in itself doesn't know about the new `Boo` instance. During intra method analysis, method A walks through the points-to graph of `x` which then sees it has been passed to `B`, method B lets A know that the parameter escape globally, hence method A makes the new instance of `Boo` global escape too.
2. Then, when a method analysis completes in inter-method analysis, for any references that are global escape, objects referenced also become global escape. For example, imagine an `New` instance reference gets assigned to a static variable. During intra analysis the static variable handle would be marked as global escape. During inter method analysis, the `New` instance becomes global escape too. This happens during inter method analysis because caller method escape value can switch to global escape, e.g. if a callee method assigns an argument to a static field.

Reply:
This map could probably be made slightly more typesafe by making it a `Map<ValueHandle, Node>`.

Reply:
Could you put this information into the code as comments?